---
title: 'STAT 555 : RNA-Seq Analysis (Group Project)'
date: "4/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
```

3 pairs of cell lines (1=HSC, 2=CMP, 3=CFUE, 4=Erythroblast)

Compare: 24,12,23 => CMP vs Erythroblast, HSC vs CMP, CMP vs CFUE : Use ScriptSeq data

1. What genes are differentially expressed across each pair of cell lines?
2. What are the functions of the genes with differential expression patterns?
3. How consistent are the results between DEseq2 and limma voom?
4. Construct a hierarchical tree using all the RNA-seq data, and perform clustering analysis. Describe the relationship based on your results.

#wget https://bioconductor.org/packages/3.12/data/experiment/src/contrib/tximportData_1.18.0.tar.gz 

### Read Data files

```{r}
library(tximport)
dir <- "../data/rna_seq/"
coldata <- read.csv("encode_ids.csv")
samples <- coldata$Sample
files <- paste0(dir, samples, ".tsv")
names(files) <- samples
```


```{r}
#Read RSEM data
txi.rsem <- tximport(files, type = "rsem", txIn = FALSE, txOut = FALSE)
gene_counts <- txi.rsem$counts

#filter out low count genes 
keep <- rowSums(gene_counts)>=10
gene_counts <- gene_counts[keep,]
```


### Gene Annotation
```{r}
library(dplyr)
library(biomaRt)

#Add genes to count_data
count_data <- as.data.frame(gene_counts) %>% mutate_if(is.numeric, round)
count_data$Gene <- sapply(strsplit(as.character(row.names(gene_counts)), '\\.'), '[[', 1)

#Assign symbols with biomart
mart <- useMart(biomart='ENSEMBL_MART_ENSEMBL',dataset='mmusculus_gene_ensembl')
genes <- biomaRt::getBM(attribute=c('ensembl_gene_id', 'mgi_symbol'), values=count_data$Gene, mart=mart,useCache = FALSE)
genes <- genes[!duplicated(genes$ensembl_gene_id),]

#Merge with count data
count_data <- merge(count_data, genes, by.x="Gene", by.y="ensembl_gene_id") %>% dplyr::rename("gene_symbol"="mgi_symbol")
row.names(count_data) <- count_data$Gene
count_data$gene_symbol[which(count_data$gene_symbol=='')] <- NA

```



### CMP vs Erythroblast cells

```{r}
#Assign variables
cell1 <- "CMP"
cell2 <- "Erythroblast"

#Filter coldata by conditions
coldata_curr <- coldata[which(coldata$Status==cell1 | coldata$Status==cell2),]
coldata_curr <- coldata_curr[order(as.character(coldata_curr$Sample)), ]
coldata_curr$Status <- factor(coldata_curr$Status, levels = c(cell1,cell2)) 
rownames(coldata_curr) <- coldata_curr$Sample
coldata_curr$Sample <- NULL
```

### DE analysis using DESEQ2

```{r, eval=T, echo=F, message=F}

#subset countdata and pdata to 
countdata_curr <- count_data[,which(colnames(count_data) %in% rownames(coldata_curr))] 

#columns must be ordered in the same way
countdata_curr <- countdata_curr[ ,order(as.character(names(countdata_curr)))] 

library(DESeq2)
ddsFullCountTable <- DESeqDataSetFromMatrix(countData = countdata_curr, colData = coldata_curr, design = ~ Status)
dds <- DESeq(ddsFullCountTable)
#normalized counts
dds_count <- counts(dds, normalize=T)
```

```{r}
# output results
res <- results(dds, cooksCutoff=FALSE, independentFiltering=FALSE)
res$gene_id <- row.names(res) 

# add gene symbols for DE results
res_df <- as.data.frame(res)
res_df <- merge(count_data[which(count_data$Gene%in%res_df$gene_id),c("Gene","gene_symbol")],res_df,by.x="Gene",by.y="gene_id")
res_df <- res_df[order(res_df$pvalue, decreasing=FALSE),]
```

### MA plot

```{r, eval=T, echo=F}
plotMA(dds,ylim=c(-2,2))
```

### Distribution of adjusted p-values

```{r, eval=T, echo=F}
hist(res$padj, main=" ", xlab="adjusted p-value", breaks=50, col="deepskyblue4", xaxp  = c(0, 1, 20))
```

### TOP20 genes Heatmap

```{r fig.height=10, fig.width=12}
library(gplots)
library(viridis)

#assign colors
colour_status <- c("navy","red")
names(colour_status) <- c(cell1,cell2) 
colour_status_list <- unlist(lapply(coldata_curr$Status,function(x){colour_status[x]}))

make_heatmap <- function(df){
  #Log2 of counts
  mat = log2(dds_count+1)
  
  #Top 20 genes
  top.mat <- mat[rownames(mat)%in%df[1:50,"Gene"],] 
  array_name <- colnames(mat)
  gene_symbol=df[1:50,"gene_symbol"]
  
  #Make heatmap
  heatmap.2(na.omit(top.mat), col=viridis(256, option="B"),
    ColSideColors=colour_status_list, #use list of assigned colors
    labCol=array_name, labRow=gene_symbol, 
    trace="none",
    margins=c(12,20), 
    cexRow=1,cexCol=1,
    keysize=1.5,key.title=NA,key.xlab="Log2-normalized counts",key.ylab="Counts",
    main="")
  legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.8) #use list of assigned colors
}

make_heatmap(res_df)
```
### All gene heatmap
make_heatmap <- function(df){
  #Log2 of counts
  mat = log2(dds_count+1)

#Include all genes
top.mat <- mat[rownames(mat)%in%df[1:18348,"Gene"],] 
array_name <- colnames(mat)
gene_symbol=df[1:18348,"gene_symbol"]

#Make heatmap
heatmap.2(na.omit(top.mat), col=viridis(256, option="B"),
          ColSideColors=colour_status_list, #use list of assigned colors
          labCol=array_name, labRow=FALSE, 
          trace="none",
          margins=c(12,20), 
          cexRow=1,cexCol=1,
          keysize=1.5,key.title=NA,key.xlab="Log2-normalized counts",key.ylab="Counts",
          main="")
legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.8) #use list of assigned colors
}
make_heatmap(res_df)
```

### Volcano Plot FOR DESEQ2

```{r}
library(ggplot2)
library(readr)
library(ggrepel)
library(extrafont)

#Adding additional columns to res_df test for significance and log2FC threshold
res_df$Expression <- "NO"
res_df$Expression[res_df$log2FoldChange > 1 & res_df$pvalue < 0.05] <- "UP"
res_df$Expression[res_df$log2FoldChange < -1 & res_df$pvalue < 0.05] <- "DOWN"
res_df$label <- NA
res_df$label[res_df$Expression != "NO"] <- res_df$gene_symbol[res_df$Expression != "NO"]

volcanoplot <- ggplot(res_df, aes(x=log2FoldChange, y=-log10(pvalue), col = Expression, label = label))+
  geom_point()+
  theme_classic()+
  theme(text = element_text(size = 20, family = "Arial", color = "black"))+
  geom_vline(xintercept=c(-1, 1), col="grey", linetype = "dashed") +
  geom_hline(yintercept=-log10(0.05), col="grey", linetype = "dashed")+
  scale_color_manual(values = c("blue", "grey", "red"))+
  xlim(-4, 4)+
  geom_text_repel(aes(fontface = "bold"))+
  theme(legend.position = "top")+
  labs(x=expression(log[2](Fold~change)), y=expression(-log[10](p-value)))
volcanoplot
```

```{r}
go_enrichment <- function(df) {

  # we want the log2 fold change 
  original_gene_list <- df$log2FoldChange
  
  # name the vector
  names(original_gene_list) <- df$gene_id
  
  # omit any NA values 
  gene_list<-na.omit(original_gene_list)
  
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  
  gse <- gseGO(geneList=gene_list, 
               keyType = "ENSEMBL", 
               nPerm = 10000, 
               minGSSize = 3, 
               maxGSSize = 800, 
               pvalueCutoff = 0.05, 
               verbose = TRUE, 
               OrgDb = org.Mm.eg.db, 
               pAdjustMethod = "none")
  return(gse)
}
```

```{r}
CMP_vs_ETH <- go_enrichment(res)

png(file="CMP vs ETH_go_rna_seq.png",
width=5000, height=3500, res=400)
dotplot(CMP_vs_ETH, showCategory=10, split=".sign",
        title="CMP vs ETH") + facet_grid(.~.sign)
dev.off()
```
_____

### DE analysis using limma voom

```{r}
library(edgeR)
status_df <- coldata %>% dplyr::filter(Sample %in% colnames(countdata_curr))
status_df <- status_df[match(names(countdata_curr), status_df$Sample),]
celltypes <- as.vector(status_df$Status)

#Create DEGlist object
deg <- DGEList(countdata_curr)

#Get normalization factors
td <- calcNormFactors(deg)
```

### Voom transformation, calculation of variance

```{r}
modmat <- model.matrix(~0 + celltypes)
mv <- voom(td, modmat, plot = T)
```

### Fitting a linear model

```{r}
fitmod <- lmFit(mv, modmat)
head(coef(fitmod))
```

### Make contrasts

```{r}
contrast <- makeContrasts(celltypesCMP - celltypesErythroblast, levels = colnames(coef(fitmod)))

#estimate contrasts for each gene and apply eBayes
estcont <- eBayes(contrasts.fit(fitmod, contrast))

#See diff expression results
top.table <- topTable(estcont, sort.by = "P", n = Inf)
top.table$Gene <- row.names(top.table)

#Add gene symbol
res_vdf <- merge(top.table, genes, by.x="Gene", by.y="ensembl_gene_id") %>% dplyr::rename("gene_symbol"="mgi_symbol")
row.names(res_vdf) <- res_vdf$Gene
res_vdf$gene_symbol[which(res_vdf$gene_symbol=='')] <- NA

#View results
res_vdf
```
### Heatmap for limma voom

```{r fig.height=10, fig.width=12}
make_heatmap(res_vdf)
```

### Volcano Plot FOR limma voom

```{r}
#Make sure to load previous libraries in the DESEQ2 version

#Adding additional columns to res_df test for significance and log2FC threshold
#The first command is because limma voom only gives logFC of 10 --> made a column for log2FC
res_vdf$log2foldchange <- log2(10^(res_vdf$logFC))
res_df$Expression <- "NO"
res_df$Expression[res_df$log2foldchange > 1 & res_df$P.Value < 0.05] <- "UP"
res_df$Expression[res_df$log2foldchange < -1 & res_df$P.Value < 0.05] <- "DOWN"
res_df$label <- NA
res_df$label[res_df$Expression != "NO"] <- res_df$gene_symbol[res_df$Expression != "NO"]

volcanoplot <- ggplot(res_df, aes(x=log2FoldChange, y=-log10(pvalue), col = Expression, label = label))+
  geom_point()+
  theme_classic()+
  theme(text = element_text(size = 20, family = "Arial", color = "black"))+
  geom_vline(xintercept=c(-1, 1), col="grey", linetype = "dashed") +
  geom_hline(yintercept=-log10(0.05), col="grey", linetype = "dashed")+
  scale_color_manual(values = c("blue", "grey", "red"))+
  xlim(-4, 4)+
  geom_text_repel(aes(fontface = "bold"))+
  theme(legend.position = "top")+
  labs(x=expression(log[2](Fold~change)), y=expression(-log[10](p-value)))
volcanoplot
```
### RNA-seq comparison 2: HSC vs CMP cells

```{r}
#Assign variables
cell1 <- "HSC"
cell2 <- "CMP"
#Filter coldata by conditions
coldata_curr <- coldata[which(coldata$Status==cell1 | coldata$Status==cell2),]
coldata_curr <- coldata_curr[order(as.character(coldata_curr$Sample)), ]
coldata_curr$Status <- factor(coldata_curr$Status, levels = c(cell1,cell2)) 
rownames(coldata_curr) <- coldata_curr$Sample
coldata_curr$Sample <- NULL
```

### DE analysis using DESEQ2 for comparison 2

```{r, eval=T, echo=F, message=F}
#subset countdata and pdata to 
countdata_curr <- count_data[,which(colnames(count_data) %in% rownames(coldata_curr))] 
#columns must be ordered in the same way
countdata_curr <- countdata_curr[ ,order(as.character(names(countdata_curr)))] 
library(DESeq2)
ddsFullCountTable <- DESeqDataSetFromMatrix(countData = countdata_curr, colData = coldata_curr, design = ~ Status)
dds <- DESeq(ddsFullCountTable)
#normalized counts
dds_count <- counts(dds, normalize=T)
```

```{r}
# output results
res <- results(dds, cooksCutoff=FALSE, independentFiltering=FALSE)
res$gene_id <- row.names(res) 
# add gene symbols for DE results
res_df <- as.data.frame(res)
res_df <-  merge(count_data[which(count_data$Gene%in%res_df$gene_id),c("Gene","gene_symbol")],res_df,by.x="Gene",by.y="gene_id")
res_df <- res_df[order(res_df$pvalue, decreasing=FALSE),]
# writing out a csv to look at genes in excel
write.csv(res_df, "HSC_CMP_compare.csv")
```
### Volcano Plot FOR DESEQ2 comparison 2

```{r}
#Remember to load libraries from comparison 1

#Adding additional columns to res_df test for significance and log2FC threshold
res_df$Expression <- "NO"
res_df$Expression[res_df$log2FoldChange > 1 & res_df$pvalue < 0.05] <- "UP"
res_df$Expression[res_df$log2FoldChange < -1 & res_df$pvalue < 0.05] <- "DOWN"
res_df$label <- NA
res_df$label[res_df$Expression != "NO"] <- res_df$gene_symbol[res_df$Expression != "NO"]

volcanoplot <- ggplot(res_df, aes(x=log2FoldChange, y=-log10(pvalue), col = Expression, label = label))+
  geom_point()+
  theme_classic()+
  theme(text = element_text(size = 20, family = "Arial", color = "black"))+
  geom_vline(xintercept=c(-1, 1), col="grey", linetype = "dashed") +
  geom_hline(yintercept=-log10(0.05), col="grey", linetype = "dashed")+
  scale_color_manual(values = c("blue", "grey", "red"))+
  xlim(-4, 4)+
  geom_text_repel(aes(fontface = "bold"))+
  theme(legend.position = "top")+
  labs(x=expression(log[2](Fold~change)), y=expression(-log[10](p-value)))
volcanoplot
```
### All gene heatmap for comparison 2
make_heatmap <- function(df){
#Log2 of counts
mat = log2(dds_count+1)

#All genes
top.mat <- mat[rownames(mat)%in%df[1:18348,"Gene"],] 
array_name <- colnames(mat)
gene_symbol=df[1:18348,"gene_symbol"]

#Make heatmap
heatmap.2(na.omit(top.mat), col=viridis(256, option="B"),
          ColSideColors=colour_status_list, #use list of assigned colors
          labCol=array_name, labRow=FALSE, 
          trace="none",
          margins=c(12,20), 
          cexRow=1,cexCol=1,
          keysize=1.5,key.title=NA,key.xlab="Log2-normalized counts",key.ylab="Counts",
          main="")
legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.8) #use list of assigned colors
}
make_heatmap(res_df)

### RNA-seq comparison 3: CMP vs CFU-E cells

```{r}
#Assign variables
cell1 <- "CMP"
cell2 <- "CFU-E"
#Filter coldata by conditions
coldata_curr <- coldata[which(coldata$Status==cell1 | coldata$Status==cell2),]
coldata_curr <- coldata_curr[order(as.character(coldata_curr$Sample)), ]
coldata_curr$Status <- factor(coldata_curr$Status, levels = c(cell1,cell2)) 
rownames(coldata_curr) <- coldata_curr$Sample
coldata_curr$Sample <- NULL
```

### DE analysis using DESEQ2 for comparison 3

```{r, eval=T, echo=F, message=F}
#subset countdata and pdata to 
countdata_curr <- count_data[,which(colnames(count_data) %in% rownames(coldata_curr))] 
#columns must be ordered in the same way
countdata_curr <- countdata_curr[ ,order(as.character(names(countdata_curr)))] 
library(DESeq2)
ddsFullCountTable <- DESeqDataSetFromMatrix(countData = countdata_curr, colData = coldata_curr, design = ~ Status)
dds <- DESeq(ddsFullCountTable)
#normalized counts
dds_count <- counts(dds, normalize=T)
```

```{r}
# output results
res <- results(dds, cooksCutoff=FALSE, independentFiltering=FALSE)
res$gene_id <- row.names(res) 
# add gene symbols for DE results
res_df <- as.data.frame(res)
res_df <-  merge(count_data[which(count_data$Gene%in%res_df$gene_id),c("Gene","gene_symbol")],res_df,by.x="Gene",by.y="gene_id")
res_df <- res_df[order(res_df$pvalue, decreasing=FALSE),]
# writing out a csv to look at genes in excel
write.csv(res_df, "CMP_CFUE_compare.csv")
```
### Volcano Plot FOR DESEQ2 comparison 3

```{r}
#Remember to load libraries from comparison 1

#Adding additional columns to res_df test for significance and log2FC threshold
res_df$Expression <- "NO"
res_df$Expression[res_df$log2FoldChange > 1 & res_df$pvalue < 0.05] <- "UP"
res_df$Expression[res_df$log2FoldChange < -1 & res_df$pvalue < 0.05] <- "DOWN"
res_df$label <- NA
res_df$label[res_df$Expression != "NO"] <- res_df$gene_symbol[res_df$Expression != "NO"]

volcanoplot <- ggplot(res_df, aes(x=log2FoldChange, y=-log10(pvalue), col = Expression, label = label))+
  geom_point()+
  theme_classic()+
  theme(text = element_text(size = 20, family = "Arial", color = "black"))+
  geom_vline(xintercept=c(-1, 1), col="grey", linetype = "dashed") +
  geom_hline(yintercept=-log10(0.05), col="grey", linetype = "dashed")+
  scale_color_manual(values = c("blue", "grey", "red"))+
  xlim(-4, 4)+
  geom_text_repel(aes(fontface = "bold"))+
  theme(legend.position = "top")+
  labs(x=expression(log[2](Fold~change)), y=expression(-log[10](p-value)))
volcanoplot
```
### All gene heatmap for comparison 3
make_heatmap <- function(df){
#Log2 of counts
mat = log2(dds_count+1)

#All genes
top.mat <- mat[rownames(mat)%in%df[1:18348,"Gene"],] 
array_name <- colnames(mat)
gene_symbol=df[1:18348,"gene_symbol"]

#Make heatmap
heatmap.2(na.omit(top.mat), col=viridis(256, option="B"),
          ColSideColors=colour_status_list, #use list of assigned colors
          labCol=array_name, labRow=FALSE, 
          trace="none",
          margins=c(12,20), 
          cexRow=1,cexCol=1,
          keysize=1.5,key.title=NA,key.xlab="Log2-normalized counts",key.ylab="Counts",
          main="")
legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.8) #use list of assigned colors
}
make_heatmap(res_df)

