---
title: 'STAT555 Project : ATAC-Seq Analysis'
author: "Avantika Diwadkar"
date: "4/20/2022"
output: html_document
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
```

### Libraries

```{r}
library(GenomicRanges)
library(dplyr)
library(tidyr)
library(DESeq2)
library(BSgenome.Mmusculus.UCSC.mm10)
#library(tracktables)
library(clusterProfiler)
library(ChIPseeker)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(biomaRt)
mart <- useMart(biomart='ENSEMBL_MART_ENSEMBL',dataset='mmusculus_gene_ensembl')
library(gplots)
library(viridis)
library(ggplot2)
library(DT)
```

### Read in the sample info file created 

Tutorial: https://rockefelleruniversity.github.io/RU_ATAC_Workshop.html 

3 pairs of cell lines (1=HSC, 2=CMP, 3=CFUE, 4=Erythroblast)

Compare: 24,12,23 => CMP vs Erythroblast, HSC vs CMP, CMP vs CFUE

5. What regions have differential chromatin patterns across each pair of cell lines? What are the genes near these regions?
6. How are differential chromatin patterns related to the expression patterns of nearby genes?
7. What are the functions of the genes with differential chromatin patterns?
8. Construct a hierarchical tree using all the ATAC-seq data and use clustering analysis to explore the pattern of cell-line specified genes. Do you get the same structure as the tree from RNA-seq data?

## Read in sample data 

```{r}
sample_info <- read.csv("atac_seq_sample_sheet.csv")
```


## Read in consensus (merged) peak intensity files generated with convert_atac_seq.sh script

```{r}
# List of file prefixes
fileNames <- c("ENCFF181AMY","ENCFF255IVU","ENCFF343PTQ","ENCFF599ZDJ",
"ENCFF616EWK","ENCFF662DYG","ENCFF796ZSB", "ENCFF832UUS")

# The peak identities
peakNames <- read.csv(paste0(fileNames[1], ".master_peak_list.bigbedfile1.pkidsort.tab"),  sep="\t")[,1]

# Get column 6 from each of the files
cols <- c()
 for (file in fileNames){
   cols <- cbind(cols, read.csv(paste0(file, ".master_peak_list.bigbedfile1.pkidsort.tab"),  sep="\t")[,6])
 }
```

```{r}
# Turn it into a data frame and add row names and column names
all <- data.frame(cols, row.names = peakNames)
colnames(all) <- fileNames

#Re-order columns and convert to integer
myCounts <- all[,as.vector(sample_info$Sample)] %>% mutate_if(is.numeric, round)
Group <- factor(as.vector(sample_info$Condition))

#View
head(myCounts)
```


## Differential ATAC-Seq analysis

The presense or absense of a peak does not fully capture the changes in ATAC-seq signal. Identifying changes in signal within peaks will allow us to better capture ATAC-seq signal differences across the celltypes.

### DESEQ2 Analysis

```{r}
#Set Group
metaData <- data.frame(Group, row.names = colnames(myCounts))
metaData$Group <- factor(metaData$Group, levels = c("HSC","CMP","CFUE","Erythroblast")) #set order 

atacDDS_table <- DESeqDataSetFromMatrix(myCounts, metaData, ~Group) #rowRanges = consensusToCount
atacDDS <- DESeq(atacDDS_table)

#normalized counts
atacDDS_count <- counts(atacDDS, normalize=T)
```


### PCA Plot

```{r}
atac_Rlog <- rlog(atacDDS)
plotPCA(atac_Rlog, intgroup = "Group", ntop = nrow(atac_Rlog))
```

### CMP vs Erythroblasts : comparison 1

```{r}
cell1 <- "CMP"
cell2 <- "Erythroblast"

ErythroMinusCMP <- results(atacDDS, c("Group", cell2 , cell1), format = "DataFrame") #CMP is the control base tissue similar to RNASeq
ErythroMinusCMP$peaks <- row.names(ErythroMinusCMP) 
ErythroMinusCMP <- ErythroMinusCMP[order(ErythroMinusCMP$pvalue),]
```


### Annotation of differential regions for comparison 1

```{r}
#Convert to GRanges
ErythroMinusCMPGR <- makeGRangesFromDataFrame(as.data.frame(ErythroMinusCMP) %>% separate(peaks, c("Chr", "Start", "Stop")),keep.extra.columns=TRUE)

#Annotate peaks
anno_ErythroMinusCMP <- annotatePeak(ErythroMinusCMPGR, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene)
```


### Get gene symbol for transcript ID

```{r}
anno_ErythroMinusCMP_df <- as.data.frame(as.GRanges(anno_ErythroMinusCMP))
anno_ErythroMinusCMP_df$Gene <- sapply(strsplit(as.character(anno_ErythroMinusCMP_df$transcriptId), '\\.'), '[[', 1)

#Assign symbols with biomart
#mart <- useEnsembl(biomart='ENSEMBL_MART_ENSEMBL',dataset='mmusculus_gene_ensembl',mirror = "useast")
genes <- biomaRt::getBM(attribute=c('ensembl_transcript_id', 'mgi_symbol'), values=anno_ErythroMinusCMP_df$Gene, mart=mart,useCache = FALSE)
genes <- genes[!duplicated(genes$ensembl_transcript_id),]

#Merge with count data
anno_ErythroMinusCMP_df <- merge(anno_ErythroMinusCMP_df, genes, by.x="Gene", by.y="ensembl_transcript_id") %>% dplyr::rename("gene_symbol"="mgi_symbol")
anno_ErythroMinusCMP_df$gene_symbol[which(anno_ErythroMinusCMP_df$gene_symbol=='')] <- NA
```


### Final results

```{r}
#Filter by significance, distance to TSS and FC
ErythroMinusCMP_results <- anno_ErythroMinusCMP_df %>% dplyr::filter(pvalue < 0.05, abs(distanceToTSS) < 5000, abs(log2FoldChange) > 0) %>% dplyr::arrange(pvalue)

#Add peaks
ErythroMinusCMP_results$Peak <- as.character(paste0(ErythroMinusCMP_results$seqnames,"_",ErythroMinusCMP_results$start,"_",ErythroMinusCMP_results$end))

#Save results
write.csv2(ErythroMinusCMP_results, "ErythrovsCMP_atacseq_sigpval_5kb_logFC.tab", row.names=F,quote = F, sep = "\t")
```

```{r}

```
### Heatmap 

```{r fig.height=10, fig.width=12}
make_heatmap <- function(df, coldata){
    
  #Log2 of counts
  norm_counts = log2(atacDDS_count+1)
  mat <- norm_counts[, row.names(coldata)]
  
  #Top 50 genes
  top.mat <- mat[rownames(mat)%in%df[1:20,"Peak"],] 
  array_name <- colnames(mat)
  gene_symbol=df[1:20,"gene_symbol"]
  
  #assign colors
  colour_status <- c("navy","red")
  names(colour_status) <- as.vector(unique(coldata$Group))
  colour_status_list <- unlist(lapply(as.vector(coldata$Group),function(x){colour_status[x]}))
  
  #Make heatmap
  heatmap.2(na.omit(top.mat), col=viridis(256, option="B"),
    ColSideColors=colour_status_list, #use list of assigned colors
    labCol=array_name, labRow=gene_symbol, 
    trace="none",
    margins=c(12,20), 
    cexRow=1,
    cexCol=1,
    keysize=1.5,key.title=NA,key.xlab="Log2-normalized counts",key.ylab="Counts",
    main="")
  legend("bottomleft",legend=names(colour_status),fill=colour_status,cex=0.8) #use list of assigned colors
}

```


```{r fig.height=8, fig.width=8}
#Filter data for cell types
coldata <- metaData %>% dplyr::filter(Group %in% c(cell1,cell2))
make_heatmap(ErythroMinusCMP_results, coldata)
```
### GO Enrichment

```{r}
go_enrichment <- function(df) {

  # we want the log2 fold change 
  original_gene_list <- df$log2FoldChange
  
  # name the vector
  names(original_gene_list) <- df$geneId
  
  # omit any NA values 
  gene_list<-na.omit(original_gene_list)
  
  # sort the list in decreasing order (required for clusterProfiler)
  gene_list = sort(gene_list, decreasing = TRUE)
  
  gse <- gseGO(geneList=gene_list, 
               keyType = "ENTREZID", 
               nPerm = 10000, 
               minGSSize = 3, 
               maxGSSize = 800, 
               pvalueCutoff = 0.05, 
               verbose = TRUE, 
               OrgDb = org.Mm.eg.db, 
               pAdjustMethod = "none")
  return(gse)
}
```
```{r}
CMP_vs_Erythroblasts <- go_enrichment(ErythroMinusCMP_results)
```
```{r}
png(file="CMP_vs_Erythroblasts_go.png",
width=5000, height=3500, res=400)
dotplot(CMP_vs_Erythroblasts, showCategory=10, split=".sign",
        title="CMP vs Erythroblasts") + facet_grid(.~.sign)
dev.off()
```
```{r}
ridgeplot(gse) + labs(x = "enrichment distribution")
```

____

## HSC vs CMP : comparison 2

```{r}
cell1 <- "HSC"
cell2 <- "CMP"

CMPMinusHSC <- results(atacDDS, c("Group", cell2 , cell1), format = "DataFrame") #HSC is control baseline
CMPMinusHSC$peaks <- row.names(CMPMinusHSC) 
CMPMinusHSC <- CMPMinusHSC[order(CMPMinusHSC$pvalue),]
```


### Annotation of differential regions for comparison 2

```{r}
#Convert to GRanges
CMPMinusHSCGR <- makeGRangesFromDataFrame(as.data.frame(CMPMinusHSC) %>% separate(peaks, c("Chr", "Start", "Stop")),keep.extra.columns=TRUE)

#Annotate peaks
anno_CMPMinusHSC <- annotatePeak(CMPMinusHSCGR, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene)
```


### Get gene symbol for transcript ID

```{r}
anno_CMPMinusHSC_df <- as.data.frame(as.GRanges(anno_CMPMinusHSC))
anno_CMPMinusHSC_df$Gene <- sapply(strsplit(as.character(anno_CMPMinusHSC_df$transcriptId), '\\.'), '[[', 1)

#Assign symbols with biomart
#mart <- useEnsembl(biomart='ENSEMBL_MART_ENSEMBL',dataset='mmusculus_gene_ensembl',mirror = "useast")
genes <- biomaRt::getBM(attribute=c('ensembl_transcript_id', 'mgi_symbol'), values=anno_CMPMinusHSC_df$Gene, mart=mart,useCache = FALSE)
genes <- genes[!duplicated(genes$ensembl_transcript_id),]

#Merge with count data
anno_CMPMinusHSC_df <- merge(anno_CMPMinusHSC_df, genes, by.x="Gene", by.y="ensembl_transcript_id") %>% dplyr::rename("gene_symbol"="mgi_symbol")
anno_CMPMinusHSC_df$gene_symbol[which(anno_CMPMinusHSC_df$gene_symbol=='')] <- NA
```


### Final results

```{r}
#Filter by significance, distance to TSS and FC
CMPMinusHSC_results <- anno_CMPMinusHSC_df %>% dplyr::filter(pvalue < 0.05, abs(distanceToTSS) < 5000, abs(log2FoldChange) > 0) %>% dplyr::arrange(pvalue)

#Add peaks
CMPMinusHSC_results$Peak <- as.character(paste0(CMPMinusHSC_results$seqnames,"_",CMPMinusHSC_results$start,"_",CMPMinusHSC_results$end))

#Save results
write.csv2(CMPMinusHSC_results, "CMPvsHSC_atacseq_sigpval_5kb_logFC.tab", row.names=F,quote = F,
          sep="\t")

```

### Heatmap 

```{r fig.height=8, fig.width=8}
#Filter data for cell types
coldata <- metaData %>% dplyr::filter(Group %in% c(cell1,cell2))
make_heatmap(CMPMinusHSC_results, coldata)
```


### GO Enrichment
```{r}
HSC_vs_CMP <- go_enrichment(CMPMinusHSC_results)

png(file="HSC_vs_CMP_go.png",
width=5000, height=3500, res=400)
dotplot(HSC_vs_CMP, showCategory=10, split=".sign",
        title="HSC vs CMP") + facet_grid(.~.sign)
dev.off()
```
____

### CMP vs CFUE : comparison 3

```{r}
cell1 <- "CMP"
cell2 <- "CFUE"

CFUEMinusCMP <- results(atacDDS, c("Group", cell2 , cell1), format = "DataFrame") #CMP is control baseline
CFUEMinusCMP$peaks <- row.names(CFUEMinusCMP) 
CFUEMinusCMP <- CFUEMinusCMP[order(CFUEMinusCMP$pvalue),]
```


### Annotation of differential regions for comparison 2

```{r}
#Convert to GRanges
CFUEMinusCMPGR <- makeGRangesFromDataFrame(as.data.frame(CFUEMinusCMP) %>% separate(peaks, c("Chr", "Start", "Stop")),keep.extra.columns=TRUE)

#Annotate peaks
anno_CFUEMinusCMP <- annotatePeak(CFUEMinusCMPGR, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene)
```


### Get gene symbol for transcript ID

```{r}
anno_CFUEMinusCMP_df <- as.data.frame(as.GRanges(anno_CFUEMinusCMP))
anno_CFUEMinusCMP_df$Gene <- sapply(strsplit(as.character(anno_CFUEMinusCMP_df$transcriptId), '\\.'), '[[', 1)

#Assign symbols with biomart
#mart <- useEnsembl(biomart='ENSEMBL_MART_ENSEMBL',dataset='mmusculus_gene_ensembl',mirror = "useast")
genes <- biomaRt::getBM(attribute=c('ensembl_transcript_id', 'mgi_symbol'), values=anno_CFUEMinusCMP_df$Gene, mart=mart,useCache = FALSE)
genes <- genes[!duplicated(genes$ensembl_transcript_id),]

#Merge with count data
anno_CFUEMinusCMP_df <- merge(anno_CFUEMinusCMP_df, genes, by.x="Gene", by.y="ensembl_transcript_id") %>% dplyr::rename("gene_symbol"="mgi_symbol")
anno_CFUEMinusCMP_df$gene_symbol[which(anno_CFUEMinusCMP_df$gene_symbol=='')] <- NA
```


### Final results

```{r}
#Filter by significance, distance to TSS and FC
CFUEMinusCMP_results <- anno_CFUEMinusCMP_df %>% dplyr::filter(pvalue < 0.05, abs(distanceToTSS) < 5000, abs(log2FoldChange) > 0) %>% dplyr::arrange(pvalue)

#Add peaks
CFUEMinusCMP_results$Peak <- as.character(paste0(CFUEMinusCMP_results$seqnames,"_",CFUEMinusCMP_results$start,"_",CFUEMinusCMP_results$end))

#Save results
write.csv2(CFUEMinusCMP_results, "CFUEvsCMP_atacseq_sigpval_5kb_logFC.tab", row.names=F,quote = F, sep="\t")

```

## Heatmap 

```{r fig.height=8, fig.width=8}
#Filter data for cell types
coldata <- metaData %>% dplyr::filter(Group %in% c(cell1,cell2))
make_heatmap(CFUEMinusCMP_results, coldata)
```

### GO Enrichment
```{r}
CMP_vs_CFUE <- go_enrichment(CFUEMinusCMP_results)

png(file="CMP_vs_CFUE_go.png",
width=5000, height=3500, res=400)
dotplot(CMP_vs_CFUE, showCategory=10, split=".sign",
        title="CMP vs CFUE") + facet_grid(.~.sign)
dev.off()
```

```{r}
eth_cmp <- c()

for (i in ErythroMinusCMP_results$annotation[ErythroMinusCMP_results$pvalue < 0.05]){
  eth_cmp <- c(eth_cmp,strsplit(i,split="[()]")[[1]][1])
}

hsc_vs_hsc <- c()
for (i in CMPMinusHSC_results$annotation[CMPMinusHSC_results$pvalue < 0.05]){
  hsc_vs_hsc <- c(hsc_vs_hsc,strsplit(i,split="[()]")[[1]][1])
}

cfue_vs_cmp <- c()
for (i in CFUEMinusCMP_results$annotation[CFUEMinusCMP_results$pvalue < 0.05]){
  cfue_vs_cmp <- c(cfue_vs_cmp,strsplit(i,split="[()]")[[1]][1])
}

stuff <- rbind(table(cfue_vs_cmp), table(hsc_vs_hsc))
stuff <- rbind(table(eth_cmp), stuff)

png(file="atac_regions_go.png",
width=3000, height=3000, res=400)
barplot(stuff, beside=TRUE, col = c("Red", "blue", "black"))
legend("topleft", legend=c("CMP vs Erythroblasts", "CMP vs CFUE", "HSC vs CMP"),
        fill = c("Red", "blue", "black"), cex=0.8)
dev.off()
```

___

### Clustering Analysis

```{r}
cmp1 <- read.csv("../integration/cmp_ery.csv")
g1 <- intersect(cmp1$RNA, cmp1$ATAC)

sg1 <- ErythroMinusCMP_results %>% dplyr::filter(gene_symbol %in% g1, pvalue < 0.01)

cmp2 <- read.csv("../integration/hsc_cmp.csv")
g2 <- intersect(cmp2$RNA, cmp2$ATAC)


cmp3 <- read.csv("../integration/cmp_cfue.csv")
g3 <- intersect(cmp3$RNA, cmp3$ATAC)

overlap_genes <- unique(c(sg1$gene_symbol, g2, g3))

```


```{r fig.width=5, fig.height=5}
make_heatmap_all <- function(df, coldata){
    
  #Log2 of counts
  norm_counts = log2(atacDDS_count+1)
  mat <- norm_counts[, row.names(coldata)]
  
  #Overlapping genes
  top.mat <- mat[rownames(mat)%in% df$Peak,] 
  top.mat <- top.mat[order(match(rownames(top.mat),df$Peak)), , drop = FALSE]
  rownames(top.mat) <- df$gene_symbol
  
  #assign colors
  colour_status <- c("#1B9E77","#D95F02", "#7570B3", "#E7298A")
  names(colour_status) <- as.vector(unique(coldata$Group))
  colour_status_list <- unlist(lapply(as.vector(coldata$Group),function(x){colour_status[x]}))
  
  #Make heatmap
  pheatmap(na.omit(top.mat), scale = "row", clustering_distance_rows = "euclidean", annotation_col = coldata)
}

# Get info sample data
coldata <- metaData 

#Get peaks for overlap genes
celresults <- rbind(rbind(CMPMinusHSC_results, CFUEMinusCMP_results), ErythroMinusCMP_results) %>% dplyr::filter(gene_symbol %in% overlap_genes) %>%
  #Keep most significant region for multiple gene annotations
  dplyr::group_by(gene_symbol) %>% dplyr::slice(which.min(pvalue)) %>% 
  dplyr::select(Peak, gene_symbol) %>% unique()

#View results
make_heatmap_all(celresults, coldata)
```

